### PR TITLE
sbotcli: add subset cmd (getReplication.getSubset)

### PR DESCRIPTION
**Pull request by cblgh**

### CODE REVIEW

**Comment by cblgh** [comment]
here we go again!!

**Comment by cryptix** [comment]
lmk if you want feedback on specifics here

**Comment by cblgh** [comment]
@cryptix yeah shoot! i'll be looping back to this code in the afternoon, already spotted at least one buggy bit but feel free to point out them all 🖤

File: cmd/sbotcli/main.go [hunk 1]
```diff
@@ -307,6 +308,65 @@ var sourceCmd = &cli.Command{
...
 }
 
+var getSubsetCmd = &cli.Command{
+	Name:  "subset",
+	Usage: "invoke the partialReplication.getSubset muxrpc",
+
+	// define cli flags
+	Flags: []cli.Flag{
+		&cli.IntFlag{Name: "limit", Value: -1},
+		&cli.BoolFlag{Name: "dsc", Value: false},
```
Comment (cryptix) [inline cmd/sbotcli/main.go:318]:
```suggestion
		&cli.BoolFlag{Name: "desc", Value: false, Usage: "order results in descending order? default is ascending"},
```

File: cmd/sbotcli/main.go [hunk 2]
```diff
@@ -307,6 +308,65 @@ var sourceCmd = &cli.Command{
...
+			query.SubsetOptions
+		}
+
+		err = json.Unmarshal([]byte(input), &payload)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal input (%w)", err)
+		}
+		// set the options after unmarshaling the input into struct form
+		payload.PageLimit = ctx.Int("limit")
+		payload.Descending = ctx.Bool("dsc")
```
Comment (cryptix) [inline cmd/sbotcli/main.go:352]:
```suggestion
		payload.Descending = ctx.Bool("desc")
```

File: cmd/sbotcli/main.go [hunk 3]
```diff
@@ -307,6 +308,65 @@ var sourceCmd = &cli.Command{
...
+		// a helper struct borrowed from the query package for converting the string input into a something we can send over
+		// the wire & which will be unpacked correctly
+		var payload struct {
+			Operation string `json:"op"`
+
+			Args   []query.SubsetOperation `json:"args,omitempty"`
+			String string            `json:"string,omitempty"`
+			Feed   *refs.FeedRef     `json:"feed,omitempty"`
+			// embed query.SubsetOptions (adds limit, dsc and keys)
+			query.SubsetOptions
```
Comment (cryptix) [inline cmd/sbotcli/main.go:343]: I think this needs to be a 2nd argument when calling the rpc.

i.e. `[{query....}, {options...}]`

**Review by cryptix — COMMENTED** [review]
a bit strapped for time so excuse my brevity/

File: cmd/sbotcli/main.go (hunk 3 above)
Comment (cblgh) [reply cmd/sbotcli/main.go:343]: ah! i was looking everywhere for the function declaration of the source call and had the hardest time finding it; thanks for the tip!

is this it? or is there an interface somewhere that i'm missing out on https://github.com/cryptoscope/go-muxrpc/blob/master/stream2_source.go

File: cmd/sbotcli/main.go [hunk 4]
```diff
@@ -307,6 +308,65 @@ var sourceCmd = &cli.Command{
...
+
+		client, err := newClient(ctx)
+		if err != nil {
+			return err
+		}
+
+		// a helper struct borrowed from the query package for converting the string input into a something we can send over
+		// the wire & which will be unpacked correctly
+		var payload struct {
+			Operation string `json:"op"`
```
[the remark this answers is not in the document]
Comment (cblgh) [reply cmd/sbotcli/main.go:337]: looks like you _might_ be wrong based on some quick tests, keeping it for now until i uncover that it's not needed! appreciate the mention

File: cmd/sbotcli/main.go (hunk 3 above)
Comment (cryptix) [reply cmd/sbotcli/main.go]: interface for..?

The arguments for a call are handled equally for all types of a call (async, duplex, source, sink). In go-land they end up on the [RawArgs](https://github.com/cryptoscope/ssb/search?q=RawArgs) field of the Request struct. Which RPC expects which count and format of the arguments is handled individually by each handler method.

File: cmd/sbotcli/blobs.go [hunk 5]
```diff
@@ -62,12 +62,13 @@ var blobsHasCmd = &cli.Command{
...
 		// TODO: direct blobstore mode!?
-		var has bool
+		var arr []bool
 		//sz, err := blobsStore.Size()
-		err = client.Async(longctx, &has, muxrpc.TypeJSON, muxrpc.Method{"blobs", "has"}, ref)
+		err = client.Async(longctx, &arr, muxrpc.TypeJSON, muxrpc.Method{"blobs", "has"}, ref)
 		if err != nil {
 			return fmt.Errorf("connect: async call failed: %w", err)
 		}
+		has := arr[0]
```
Comment (cryptix) [inline cmd/sbotcli/blobs.go:71]: Please check the length before accessing to protect against a panic.

Some implementation might emit an empty array.

File: cmd/sbotcli/simple_test.go [hunk 6]
```diff
@@ -99,6 +100,66 @@ func TestWhoami(t *testing.T) {
 	r.NoError(<-errc)
 }
 
+func TestGetSubset(t *testing.T) {
```
Comment (cryptix) [inline cmd/sbotcli/simple_test.go:103]: 👏

**Comment by cblgh** [comment]
@cryptix is this PR clear to land? ✈

**Comment by cblgh** [comment]
🎉 🎉

**Pull request merged**